demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "jfGlh7oY6lpx"
+      },
+      "source": [
+        "# How to optimize a quantum machine learning model using Catalyst and Optax\n",
+        "\n",
+        "Once you have set up your quantum machine learning model (which typically includes deciding on your circuit architecture/ansatz, determining how you embed or integrate your data, and creating your cost function to minimize a quantity of interest), the next step is **optimization**. That is, setting up a classical optimization loop to find a minimal value of your cost function.\n",
+        "\n",
+        "In this example, we'll show you how to use [JAX](https://jax.readthedocs.io), an autodifferentiable machine learning framework, and [Optax](https://optax.readthedocs.io/), a suite of JAX-compatible gradient-based optimizers, to optimize a PennyLane quantum machine learning model.\n",
+        "\n",
+        "## Set up your model, data, and cost\n",
+        "\n",
+        "Here, we will create a simple QML model for our optimization. In particular:\n",
+        "\n",
+        "- We will embed our data through a series of rotation gates.\n",
+        "- We will then have an ansatz of trainable rotation gates with parameters `weights`; it is these values we will train to minimize our cost function.\n",
+        "- We will train the QML model on `data`, a `(5, 4)` array, and optimize the model to match target predictions given by `target`."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 3,
+      "metadata": {
+        "id": "iKtDMXjENR7X"
+      },
+      "outputs": [],
+      "source": [
+        "import pennylane as qml\n",
+        "from jax import numpy as jnp\n",
+        "import jax\n",
+        "import optax\n",
+        "import catalyst"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 4,
+      "metadata": {
+        "id": "tiK-itpFNVwE"
+      },
+      "outputs": [],
+      "source": [
+        "n_wires = 5\n",
+        "data = jnp.sin(jnp.mgrid[-2:2:0.2].reshape(n_wires, -1)) ** 3\n",
+        "targets = jnp.array([-0.2, 0.4, 0.35, 0.2])\n",
+        "\n",
+        "dev = qml.device(\"lightning.qubit\", wires=n_wires)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "metadata": {
+        "id": "w4QPSXdysqJx"
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def data_embedding(i):\n",
+        "        qml.RY(data[i], wires=i)\n",
```

Comment:
Is worth presenting some built-in PennyLane embeddings, or did you chose this specifically for the control flow? :)

Reply:
I chose this to match https://pennylane.ai/qml/demos/tutorial_How_to_optimize_QML_model_using_JAX_and_Optax/, although I can't remember why we chose it there 😅

Reply:
Got it 😅